crates/biome_js_type_info/src/resolver.rs

```diff
@@ -17,81 +62,360 @@ const MAX_RESOLUTION_DEPTH: usize = 16;
 /// also have two layers of type *resolution*, both of which implement this
 /// trait.
 pub trait TypeResolver {
-    /// Resolves a type by its `qualifier`.
-    fn resolve_qualifier(&self, qualifier: &TypeReferenceQualifier) -> Option<Type>;
+    /// Returns the resolution level handled by this resolver.
+    fn level(&self) -> TypeResolverLevel;
 
-    /// Resolves the type of a value by its `identifier`.
-    fn resolve_type_of(&self, identifier: &Text) -> Option<Type>;
-}
+    /// Attempts to find the given type data if it already exists within the
+    /// resolver.
+    fn find_type(&self, type_data: &TypeData) -> Option<TypeId>;
 
-impl Type {
-    /// Attempts to resolve all unresolved types using the given resolver.
-    pub fn resolve(&mut self, resolver: &dyn TypeResolver) {
-        let stack = &[&**self];
-        *self = if self.needs_resolving(resolver, stack) {
-            self.resolved(resolver, stack)
-        } else {
-            self.flattened(resolver, stack)
-        };
+    /// Returns type data by its ID.
+    ///
+    /// **Warning:** The given [`TypeId`] must be a type ID that was created by
+    ///              this resolver.
+    fn get_by_id(&self, id: TypeId) -> &TypeData;
```

Comment:
```suggestion
    /// Returns type data by its ID.
    ///
    /// ## Warning
    /// The given [`TypeId`] must be a type ID that was created by
    /// this resolver.
    fn get_by_id(&self, id: TypeId) -> &TypeData;
```

I wonder if there's a way to prevent the mistake. Like, how do we know if the `id` passed is indeed the one created by this resolver?

Reply:
So far, the mistake hasn't been made yet, I merely put it there out of caution :sweat_smile: 

I actually find that the way the resolvers are set up, it's not super likely you'll make a mistake with this, because mostly it gets turned into `ResolvedTypeId` anyway. But yeah, if it turns out it does become a pitfall, I'd happily see if we can do something about it. For now I think I'd prefer to keep it simple with a warning :)